### DISCUSSION TITLE
RabbitMQ connection closes when running the imagemagick_files example

### DISCUSSION THREAD

**Discussion by wigging** [ORNL/Zambeze#131] [Q&A] [answered]
I'm trying to run the `imagemagick_files.py` example from the `dev-cleanup` branch. I have RabbitMQ running via a Docker container as discussed at https://www.rabbitmq.com/download.html. When I run the example I get the terminal output shown below.

```
[Zambeze Agent] [INFO] 2024-01-26 14:54:47,072 - [activities/shell.py] Printing files after init in SHELL
[Zambeze Agent] [INFO] 2024-01-26 14:54:47,072 - ['local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/01.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/02.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/03.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/04.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/05.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/06.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/07.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/08.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/09.jpg', 'local:///Users/g72/Desktop/ORNL/FlowceptZambeze/zambeze/examples/../tests/campaigns/imagesequence/10.jpg']
[Zambeze Agent] [DEBUG] 2024-01-26 14:54:47,072 - Adding activity: ImageMagick
[Zambeze Agent] [INFO] 2024-01-26 14:54:47,072 - Number of activities to dispatch: 1
ACTIVITY NAME: ImageMagick
<zambeze.campaign.activities.shell.ShellActivity object at 0x7f88f8732500>
[Zambeze Agent] [DEBUG] 2024-01-26 14:54:47,089 - Shipping activity DAG of 3 nodes...
[Zambeze Agent] [DEBUG] 2024-01-26 14:54:47,089 - Sending activity DAG via ZMQ...
[Zambeze Agent] [DEBUG] 2024-01-26 14:54:47,089 - Activity DAG successfully sent!
```

Meanwhile, in another terminal session, RabbitMQ continuously closes the connections as shown here:

```
19:56:13 [info] <0.3781.0> accepting AMQP connection <0.3781.0> (192.168.65.1:23490 -> 172.17.0.2:5672)
19:56:13 [error] <0.3781.0> closing AMQP connection <0.3781.0> (192.168.65.1:23490 -> 172.17.0.2:5672):
19:56:13 [error] <0.3781.0> {bad_header,<<255,0,0,0,0,0,0,0>>}
19:56:13 [info] <0.3786.0> accepting AMQP connection <0.3786.0> (192.168.65.1:23491 -> 172.17.0.2:5672)
19:56:13 [error] <0.3786.0> closing AMQP connection <0.3786.0> (192.168.65.1:23491 -> 172.17.0.2:5672):
19:56:13 [error] <0.3786.0> {bad_header,<<255,0,0,0,0,0,0,0>>}
19:56:13 [info] <0.3791.0> accepting AMQP connection <0.3791.0> (192.168.65.1:23492 -> 172.17.0.2:5672)
19:56:13 [error] <0.3791.0> closing AMQP connection <0.3791.0> (192.168.65.1:23492 -> 172.17.0.2:5672):
19:56:13 [error] <0.3791.0> {bad_header,<<255,0,0,0,0,0,0,0>>}
19:56:13 [info] <0.3796.0> accepting AMQP connection <0.3796.0> (192.168.65.1:23493 -> 172.17.0.2:5672)
19:56:13 [error] <0.3796.0> closing AMQP connection <0.3796.0> (192.168.65.1:23493 -> 172.17.0.2:5672):
19:56:13 [error] <0.3796.0> {bad_header,<<255,0,0,0,0,0,0,0>>}
```

I eventually have to force quit the example because RabbitMQ just keeps accepting and closing connections. Any ideas on what's going on here?

**Update:**

Running `zambeze agent start` then running the example generated the log shown below. Looks like it can't connect to the RabbitMQ server. Maybe the IP address is screwed up from the Docker container.

```
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,773 - *****************************************************
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,773 - *****************************************************
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,773 - Creating Zambeze agent subprocess with configuration:
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,773 - *****************************************************
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - Log Path:		~/.zambeze/logs/2024_01_26-16_26_31_055/zambeze.log
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - Config Path:		~/.zambeze/agent.yaml
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - Debug Logs:		True
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - ZMQ activity port:	0
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - ZMQ heartbeat port:	0
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - *****************************************************
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - *****************************************************
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,774 - 111
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,782 - Loading settings from config file: ~/.zambeze/agent.yaml
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,786 - Configuring Plugin: shell
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,788 - Configuring shell plugin
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,788 -   - action bash supported True
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,788 - Configured shell = True
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,788 - 222
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,788 - [EXECUTOR] Creating executor...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,799 - [executor] Successfully initialized Executor!
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,799 - 333
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,799 - [mh] RabbitMQ broker and channel both created successfully!
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,800 - [mh] Binding to random ZMQ port...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,802 - [mh] Advertised port in agent.yaml file: 62517
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,802 - [mh] Message handler successfully initialized!
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,802 - [recv_activities_from_campaign] Waiting for messages from campaign...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,802 - [mh] Connecting to RabbitMQ RECV ACTIVITY broker...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,802 - Preparing to start queue at address 127.0.0.1:5672
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,802 - [mh] Connecting to RabbitMQ RECV CONTROL broker...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,803 - Preparing to start queue at address 127.0.0.1:5672
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,803 - [mh] Connecting to RabbitMQ SEND ACTIVITY broker...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,803 - Preparing to start queue at address 127.0.0.1:5672
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,803 - [mh] Connecting to RabbitMQ SEND CONTROL broker...
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,803 - Preparing to start queue at address 127.0.0.1:5672
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,804 - Starting activity sorter thread!
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,804 - [agent] Starting send control thread!
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,809 - [agent] Starting control sorter thread!
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,809 - 444
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,810 - [executor] In __process! 
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,810 - [executor] Moving to working directory ~/
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,810 - [executor] Retrieving a message! 
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,816 -  SIZE OF QUEUE: 0
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,823 - 555
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,848 - Unable to connect to RabbitMQ server at 127.0.0.1:5672
1. Make sure your firewall ports are open.
2. That the rabbitmq-service is up and running.
3. The correct ip address and port have been specified.
4. That an agent.yaml file exists for the zambeze agent.

[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,865 - Unable to connect to RabbitMQ server at 127.0.0.1:5672
1. Make sure your firewall ports are open.
2. That the rabbitmq-service is up and running.
3. The correct ip address and port have been specified.
4. That an agent.yaml file exists for the zambeze agent.

[Zambeze Agent] [ERROR] 2024-01-26 16:26:31,883 - CAUGHT: <function print_exc at 0x7f99d854cee0>
[Zambeze Agent] [ERROR] 2024-01-26 16:26:31,884 - CAUGHT: <function print_exc at 0x7f99d854cee0>
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,884 - [send_control] Waiting for messages...
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,895 - Unable to connect to RabbitMQ server at 127.0.0.1:5672
1. Make sure your firewall ports are open.
2. That the rabbitmq-service is up and running.
3. The correct ip address and port have been specified.
4. That an agent.yaml file exists for the zambeze agent.

[Zambeze Agent] [ERROR] 2024-01-26 16:26:31,925 - CAUGHT: <function print_exc at 0x7f99d854cee0>
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,925 - [message_handler] [***] Waiting using persistent listener on RabbitMQ ACTIVITIES channel.
[Zambeze Agent] [INFO] 2024-01-26 16:26:31,907 - [send_activity] Waiting for messages...
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,938 - Unable to connect to RabbitMQ server at 127.0.0.1:5672
1. Make sure your firewall ports are open.
2. That the rabbitmq-service is up and running.
3. The correct ip address and port have been specified.
4. That an agent.yaml file exists for the zambeze agent.

[Zambeze Agent] [ERROR] 2024-01-26 16:26:31,973 - CAUGHT: <function print_exc at 0x7f99d854cee0>
[Zambeze Agent] [DEBUG] 2024-01-26 16:26:31,973 - [message_handler] [***] Waiting using persistent listener on RabbitMQ CONTROL channel.
```

**Comment by tskluzac** [comment]
This is interesting. I'm curious if this is a timeout issue, or if this is on the initial connection. If you do a fresh
`zambeze agent stop` 
followed by 
`zambeze agent start` 
and immediately rerun the example, do you see this same AMQP error?

**Reply by wigging** [reply]
Running `zambeze agent stop` gives an error about `ModuleNotFoundError: No module named 'flowcept'`. I thought zambeze didn't require flowcept?

**Reply by tskluzac** [reply]
It shouldn't... but it seems like it unfortunately does (I thought it skipped those lines quietly if not installed). A good find! I think the easier fix would be to just install flowcept (https://github.com/ORNL/flowcept) OR you could temporarily squash flowcept by commenting out lines 13 and 76-78 in https://github.com/ORNL/Zambeze/blob/main/zambeze/orchestration/agent/message_handler.py

**Reply by wigging** [reply]
I added the log output from running zambeze agent to my original question.

**Comment by renan-souza** [comment]
We should make flowcept optional, not required, in Zambeze but right now
there is a strong dependency in the code that is not being catched as an
exception.


pip install flowcept[zambeze] should fix this issue.

You'll have to export FLOWCEPT_SETTINGS_PATH with the path to the
settings.yaml file in the resources directory in flowcept repository

**Reply by tskluzac** [reply]
I'm wondering how to handle 'missing Flowcept' quietly. Should it be configurable in the zambeze.yaml file? Should it always *try* to import it and flush a warning to the logs if it can't be found? I'm open to either.

**Reply by renan-souza** [reply]
I'd suggest the following behavior to make FlowCept an optional dependency in Zambeze.

- Make it a flag in Zambeze's settings yaml file.
- If TRUE, an error (not warning!) should be thrown if Flowcept could not be started
- If FALSE, FlowCept APIs should not even be executed within the code, thus no exception will be thrown.

If TRUE, the user will need to install an optional dependency and set up the requirements to run FlowCept (settings.yaml and FlowCept's Docker Compose). 

The way I did in FlowCept was to use [optional dependency management](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies). 
For example, to install FlowCept's Zambeze adapter, one just needs to run `pip install flowcept[zambeze]`.

We could do the same in Zambeze, that is, if one runs `pip install zambeze[flowcept]`, it would install flowcept's adapter needed when the flag is TRUE.

**Reply by tskluzac** [reply]
Sounds good. This should be straightforward to implement as all of the changes can be captured in the message_handler. I'll get right on it.

**Comment by renan-souza** [comment]
will need flowcepts docker compose as well to run both systems together.

I think will need to change queue name in the flowcept settings.yaml

**Reply by renan-souza** [reply]
We do 3 things to run FlowCept with Zambeze:

1. Install FlowCept's Zambeze adapter: `pip install flowcept[zambeze]`
2. Run FlowCept's docker compose: https://github.com/ORNL/Zambeze/blob/dev-cleanup/deployment/compose-flowcept.yml
3. Before running an agent process, export env var as follows: `export FLOWCEPT_SETTINGS_PATH=/path/to/flowcept.yaml` A sample file of flowcept.yaml is found here: https://github.com/ORNL/flowcept/blob/main/resources/settings.yaml

Currently (only currently) there's an additional step which is to specify the queue name here: https://github.com/ORNL/flowcept/blob/main/resources/settings.yaml#L65
Soon I will update the template file to default to the actual queue name used in Zambeze.

@tskluzac what's the current queue name we use in Zambeze+FlowCept?

**Reply by tskluzac** [reply]
The names of the queue channels are "CONTROL" and "ACTIVITIES"

**Reply by renan-souza** [reply]
As of now, Zambeze's adapter in Flowcept can only intercept one single queue. 

If both queues should be intercepted by Flowcept, we can either (i) launch two different Flowcept processes, without any change in the adapter code or (ii) change the adapter code.

Or we can let Flowcept intercept one single queue in Zambeze. In this case, would it be the queue "ACTIVITIES"?

Let me know what you think we should do.

**Comment by tskluzac** [comment]
I have just pushed a 'muting' capability for flowcept that checks the log for a flowcept 'active' flag and then if false (or nonexistent) don't load any of the flowcept libraries. My config now contains this block at the top level: 

```
flowcept:
  config:
    active: false
```


Let me know if this helps!

**Reply by renan-souza** [reply]
Thanks, Tyler!

> I have just pushed a 'muting' capability for flowcept that checks the log

Not sure what you mean by checking the log. Do you mean "that checks zambeze's config file" ?

**Reply by tskluzac** [reply]
Yep, exactly.

**Comment by wigging** [comment, marked as answer]
Turns out there were several steps that I wasn't aware of to run the example. So to answer this question, I listed the steps below that successfully runs this example for me:

1. Install and run RabbitMQ. I used the Docker image to run it. See [here](https://www.rabbitmq.com/download.html).
2. If not done already, create a config file at `~/.zambeze/agent.yaml` with contents similar to the following:
```yaml
plugins:
  All:
    default_working_directory: /Users/gavin/Desktop
  shell:
    config: {}
rmq:
  host: localhost
  port: 5672
zmq:
  host: 127.0.0.1
  port: 60081
```
3. Go to the examples directory `cd zambeze/examples`
4. Run the zambeze agent using `zambeze agent start`
5. Run the example using `python imagemagick_files.py`
6. A gif image named `a.gif` should be created in the working directory defined in the `agent.yaml` file
7. Stop zambeze agent using `zambeze agent stop`
